File: src/cmds/session.ts
import { DB, readCommands } from '../core/local-db';
import { formatNumber, percent, withinDays } from '../core/utils';

interface Row {
  timestamp: string;
  cmd_type: string;
  saved_bytes: number;
  savings_pct: number;
}

export function runSession(db: DB): string {
  const rows = readCommands(db)
    .slice()
    .sort((a, b) => (a.timestamp < b.timestamp ? -1 : a.timestamp > b.timestamp ? 1 : 0)) as Row[];

  if (rows.length === 0) return 'No commands logged yet.';

  const sessions: Array<{ start: string; end: string; rows: Row[] }> = [];
  let current: Row[] = [];
  let lastTs = 0;
  for (const r of rows) {
    const ts = new Date(r.timestamp).getTime();
    if (current.length === 0) {
      current.push(r);
    } else if (ts - lastTs > 30 * 60 * 1000) {
      sessions.push({
        start: current[0].timestamp,
        end: current[current.length - 1].timestamp,
        rows: current,
      });
      current = [r];
    } else {
      current.push(r);
    }
    lastTs = ts;
  }
  if (current.length > 0) {
    sessions.push({
      start: current[0].timestamp,
      end: current[current.length - 1].timestamp,
      rows: current,
    });
  }

  const lines: string[] = ['Recent sessions:'];
  for (const s of sessions.slice(-10).reverse()) {
    const total = s.rows.length;
    const proxied = s.rows.filter((r) => r.savings_pct > 0).length;
    const saved = s.rows.reduce((acc, r) => acc + r.saved_bytes, 0);
    const startStr = formatRange(s.start, s.end);
    const pct = total > 0 ? (proxied / total) * 100 : 0;
    lines.push(
      `  ${startStr}   ${percent(pct, 0).padStart(4)} proxied (${proxied}/${total} commands)   saved ${formatNumber(saved)} bytes`,
    );
  }

  // Unproxied commands this week
  const unproxCounts = new Map<string, number>();
  for (const r of rows) {
    if (r.savings_pct === 0 && withinDays(r.timestamp, 7)) {
      unproxCounts.set(r.cmd_type, (unproxCounts.get(r.cmd_type) || 0) + 1);
    }
  }
  const unprox = Array.from(unproxCounts.entries())
    .map(([cmd_type, runs]) => ({ cmd_type, runs }))
    .sort((a, b) => b.runs - a.runs)
    .slice(0, 6);
  if (unprox.length > 0) {
    lines.push('');
    lines.push('Unproxied commands this week:');
    lines.push('  ' + unprox.map((u) => `${u.cmd_type} (×${u.runs})`).join('   '));
    lines.push('  → Run: tok discover for suggestions');
  }
  return lines.join('\n');
}

function formatRange(start: string, end: string): string {
  const s = new Date(start);
  const e = new Date(end);
  const month = s.toLocaleString('en-US', { month: 'short', day: 'numeric' });
  const sH = String(s.getHours()).padStart(2, '0') + ':' + String(s.getMinutes()).padStart(2, '0');
  const eH = String(e.getHours()).padStart(2, '0') + ':' + String(e.getMinutes()).padStart(2, '0');
  return `${month} ${sH}-${eH}`;
}

File: internal/analytics/session.go
package analytics

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/Kevalgor12/tok-proxy/internal/store"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

type sessionGroup struct {
	start string
	end   string
	rows  []store.CommandRow
}

// RunSession groups logged commands into activity sessions (a gap over 30 minutes starts a
// new one) and reports how much of each was proxied.
func RunSession(s *store.Store) string {
	rows := s.ReadCommands()
	sorted := append([]store.CommandRow(nil), rows...)
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].Timestamp < sorted[j].Timestamp })
	if len(sorted) == 0 {
		return "No commands logged yet."
	}

	var sessions []sessionGroup
	var current []store.CommandRow
	var lastT time.Time
	flush := func() {
		if len(current) == 0 {
			return
		}
		sessions = append(sessions, sessionGroup{
			start: current[0].Timestamp,
			end:   current[len(current)-1].Timestamp,
			rows:  append([]store.CommandRow(nil), current...),
		})
	}
	for _, r := range sorted {
		t, _ := parseTS(r.Timestamp)
		if len(current) == 0 {
			current = []store.CommandRow{r}
		} else if t.Sub(lastT) > 30*time.Minute {
			flush()
			current = []store.CommandRow{r}
		} else {
			current = append(current, r)
		}
		lastT = t
	}
	flush()

	lines := []string{"Recent sessions:"}
	for _, sess := range reversed(tail(sessions, 10)) {
		total := len(sess.rows)
		proxied, saved := 0, 0
		for _, r := range sess.rows {
			if r.SavingsPct > 0 {
				proxied++
			}
			saved += r.SavedBytes
		}
		pct := 0.0
		if total > 0 {
			pct = float64(proxied) / float64(total) * 100
		}
		lines = append(lines, fmt.Sprintf("  %s   %s proxied (%d/%d commands)   saved %s bytes",
			formatRange(sess.start, sess.end), util.Pad(util.Percent(pct, 0), 4, true), proxied, total, util.FormatNumber(saved)))
	}

	// Unproxied commands this week.
	unprox := newGroups[int]()
	for _, r := range sorted {
		if r.SavingsPct == 0 && util.WithinDays(r.Timestamp, 7) {
			*unprox.get(r.CmdType) += 1
		}
	}
	type row struct {
		cmdType string
		runs    int
	}
	var rows2 []row
	for _, e := range unprox.entries() {
		rows2 = append(rows2, row{e.key, *e.val})
	}
	sort.SliceStable(rows2, func(i, j int) bool { return rows2[i].runs > rows2[j].runs })
	rows2 = head(rows2, 6)
	if len(rows2) > 0 {
		var parts []string
		for _, r := range rows2 {
			parts = append(parts, fmt.Sprintf("%s (×%d)", r.cmdType, r.runs))
		}
		lines = append(lines, "", "Unproxied commands this week:", "  "+strings.Join(parts, "   "), "  → Run: tok discover for suggestions")
	}
	return strings.Join(lines, "\n")
}

func formatRange(start, end string) string {
	s, ok1 := parseTS(start)
	e, ok2 := parseTS(end)
	if !ok1 || !ok2 {
		return start
	}
	return fmt.Sprintf("%s %s-%s", s.Local().Format("Jan 2"), s.Local().Format("15:04"), e.Local().Format("15:04"))
}

func parseTS(s string) (time.Time, bool) {
	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
		return t, true
	}
	if t, err := time.Parse(time.RFC3339, s); err == nil {
		return t, true
	}
	return time.Time{}, false
}
